service/chain/live.go

```diff
@@ -0,0 +1,187 @@
+	}
+	// Empty byte array effectively subscribes to everything
+	err = psock.SetOption(mangos.OptionSubscribe, []byte("block_updates"))
+	if err != nil {
+		return nil, fmt.Errorf("could not subscribe: %w", err)
+	}
+
+	decompressor, err := zstd.NewReader(nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize decompressor: %w", err)
+	}
+
+	// rsock, err := req.NewSocket()
+	// if err != nil {
+	// 	return nil, fmt.Errorf("could not get new req socket: %w", err)
+	// }
+	// err = rsock.Dial(fmt.Sprint(host, ":", reqPort))
+	// if err != nil {
+	// 	return nil, fmt.Errorf("could not dial on req socket: %w", err)
+	// }
+
+	live := Live{
+		log: log.With().Str("component", "live-chain").Logger(),
+
+		pubsub:       psock,
+		decompressor: decompressor,
+
+		// FIXME: Switch to something that allows us to discard unnecessary elements.
+		// Could be done by deleting the entry at `height-1` when calls to `Commit`?
+		blocks: make(map[uint64]flow.Block),
+
+		done: make(chan struct{}),
+	}
+
+	return &live, nil
+}
+
+func (l *Live) Run() {
+	go func() {
+		for {
+			select {
+			case <-l.done:
+				return
+			default:
+			}
+
+			msg, err := l.pubsub.Recv()
+			if err != nil {
+				l.log.Warn().Err(err).Msg("could not receive update")
+				continue
+			}
+
+			l.log.Info().Msg("received block update")
+
+			// Get payload from message by removing topic header.
+			payload := bytes.Replace(msg, []byte("block_updates"), []byte{}, 1)
+
+			payload, err = l.decompressor.DecodeAll(payload, nil)
+			if err != nil {
+				l.log.Error().Err(err).Msg("unable to decompress block payload")
+			}
+
+			var block flow.Block
+			err = cbor.Unmarshal(payload, &block)
+			if err != nil {
+				l.log.Error().Err(err).Msg("unable to decode block payload")
+			}
+
+			// TODO: Detect if we're missing some info and request it using the req.
+
+			l.blocksMu.Lock()
+			l.blocks[block.Header.Height] = block
+			l.blocksMu.Unlock()
+		}
+	}()
+}
+
+func (l *Live) Root() (uint64, error) {
+	// Find the lowest height we have available, to use as root height.
+	var rootHeight uint64 = math.MaxUint64
+	for height := range l.blocks {
+		if height < rootHeight {
+			rootHeight = height
+		}
+	}
+
+	if rootHeight == math.MaxUint64 {
+		return 0, dps.ErrRetryLater
+	}
+
+	return rootHeight, nil
+}
+
+func (l *Live) Header(height uint64) (*flow.Header, error) {
+	l.blocksMu.RLock()
+	defer l.blocksMu.RUnlock()
+
+	block, exists := l.blocks[height]
+	if !exists {
+		return nil, dps.ErrRetryLater
+	}
+
+	return block.Header, nil
+}
+
+func (l *Live) Commit(height uint64) (flow.StateCommitment, error) {
+	l.blocksMu.RLock()
+	defer l.blocksMu.RUnlock()
+
+	block, exists := l.blocks[height]
+	if !exists {
+		return flow.StateCommitment{}, dps.ErrRetryLater
+	}
+
+	return flow.StateCommitment(block.ID()), nil
+}
+
+func (l *Live) Events(height uint64) ([]flow.Event, error) {
+	// Not implemented, but should not panic since it is used.
+	// TODO: Implement this.
```

Comment:
Yeah, this will require adding another channel to the pub socket where we push all events after a block has been executed.